File: openidm-zip/src/test/java/org/forgerock/openidm/bin/defaults/script/ConditionalRolesTest.java
/*
 * The contents of this file are subject to the terms of the Common Development and
 * Distribution License (the License). You may not use this file except in compliance with the
 * License.
 *
 * You can obtain a copy of the License at legal/CDDLv1.0.txt. See the License for the
 * specific language governing permission and limitations under the License.
 *
 * When distributing Covered Software, include this CDDL Header Notice in each file and include
 * the License file at legal/CDDLv1.0.txt. If applicable, add the following below the CDDL
 * Header, with the fields enclosed by brackets [] replaced by your own identifying
 * information: "Portions copyright [year] [name of copyright owner]".
 *
 * Copyright 2016 ForgeRock AS.
 */

package org.forgerock.openidm.bin.defaults.script;

import static org.assertj.core.api.Assertions.assertThat;
import static org.forgerock.json.JsonValue.field;
import static org.forgerock.json.JsonValue.json;
import static org.forgerock.json.JsonValue.object;

import org.forgerock.json.JsonValue;
import org.forgerock.json.resource.ResourceException;
import org.forgerock.script.Script;
import org.forgerock.script.ScriptEntry;
import org.forgerock.script.ScriptRegistry;
import org.forgerock.script.engine.ScriptEngineFactory;
import org.forgerock.script.javascript.RhinoScriptEngineFactory;
import org.forgerock.script.registry.ScriptRegistryImpl;
import org.forgerock.script.scope.Function;
import org.forgerock.script.scope.Parameter;
import org.forgerock.script.source.DirectoryContainer;
import org.forgerock.services.context.RootContext;
import org.testng.Assert;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import javax.script.ScriptException;
import java.io.IOException;
import java.net.URISyntaxException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Tests conditionalRoles.js
 */
public class ConditionalRolesTest {
    private static final String ID = "_id";
    private static final String REV = "_rev";
    private static final String MAIL = "mail";
    private static final String GIVEN_NAME = "givenName";
    private static final String SN = "sn";
    private static final String DESCRIPTION = "description";
    private static final String USER_NAME = "userName";
    private static final String TELEPHONE = "telephoneNumber";
    private static final String STATUS = "accountStatus";
    private static final String CONDITIONAL_GRANTS = "conditionalGrants";
    private static final String DIRECT_GRANTS = "directGrants";
    private static final String ROLES = "roles";
    private static final String REF = "_ref";
    private static final String REF_PROPERTIES = "_refProperties";
    private static final String GRANT_TYPE = "_grantType";
    private static final String CONDITIONAL = "conditional";
    private static final String ROLE_GRANT_ID = "566";
    private static final String NAME = "name";
    private static final String CONDITION = "condition";
    private ScriptRegistryImpl scriptRegistry;

    interface RoleState {
        String getRoleCondition();
        String getRoleDescription();
    }

    enum Role implements RoleState {
        NONE {
            public String getRoleCondition() {
                return "false";
            }

            public String getRoleDescription() {
                return "no condition";
            }
        }, MAIL_CO_COM {
            public String getRoleCondition() {
                return "/mail co 'com'";
            }

            public String getRoleDescription() {
                return "does mail contain com";
            }

        }, MAIL_CO_EXAMPLE {
            public String getRoleCondition() {
                return "/mail co 'example'";
            }

            public String getRoleDescription() {
                return "does mail contain example";
            }

        }, MAIL_CO_GOOGLE {
            public String getRoleCondition() {
                return "/mail co 'google'";
            }

            public String getRoleDescription() {
                return "does mail contain google";
            }
        }};

    enum GrantType { DIRECT, CONDITIONAL };

    @BeforeClass
    public void initScriptRegistry() throws IOException, URISyntaxException, ScriptException {
        Map<String, Object> configuration = new HashMap<>();
        configuration.put(getLanguageName(), getConfiguration());
        scriptRegistry = new ScriptRegistryImpl(configuration,
                Collections.<ScriptEngineFactory>singleton(new RhinoScriptEngineFactory()), null, null);
        URL scriptContainer = getScriptContainer("/bin/defaults/script/");
        Assert.assertNotNull(scriptContainer);
        scriptRegistry.addSourceUnit(new DirectoryContainer("bin/defaults/script", scriptContainer));
        URL lodashContainer = getScriptContainer("/scriptLibs/");
        Assert.assertNotNull(lodashContainer);
        scriptRegistry.addSourceUnit(new DirectoryContainer("scriptLibs", lodashContainer));
        Map<String, Object> console = new HashMap<>();
        console.put("log", new Function<Void>() {
            @Override
            public Void call(Parameter scope, Function<?> callback, Object... arguments) throws ResourceException, NoSuchMethodException {
                if (arguments.length > 0) {
                    if (arguments[0] instanceof String) {
                        System.out.println((String) arguments[0]);
                    }
                }
                return null;
            }
        });
        scriptRegistry.put("console", console);
    }

    protected Map<String, Object> getConfiguration() {
        Map<String, Object> configuration = new HashMap<>();
        return configuration;
    }

    protected String getLanguageName() {
        return RhinoScriptEngineFactory.LANGUAGE_NAME;
    }

    protected URL getScriptContainer(String name) {
        return ConditionalRolesTest.class.getResource(name);
    }

    public ScriptRegistry getScriptRegistry() {
        return scriptRegistry;
    }


    @DataProvider(name="UserConditionalRolesUpdate")
    Object[][] resourcePaths() {
        return new Object[][] {
                { baseUser(), systemConditionalRoles(), getUserRoleGrants(null, null) },
                { userWithRoleGrant(Role.MAIL_CO_EXAMPLE, GrantType.CONDITIONAL), systemConditionalRoles(), getUserRoleGrants(Role.MAIL_CO_EXAMPLE, null) },
                { userWithRoleGrant(Role.MAIL_CO_COM, GrantType.DIRECT), systemConditionalRoles(), getUserRoleGrants(null, Role.MAIL_CO_COM)}
        };
    }

//    @Test(dataProvider = "UserConditionalRolesUpdate")
    public void testUserConditionalRoles(JsonValue user, JsonValue existingConditionalRoles, JsonValue userRoleGrants) throws ScriptException {
        JsonValue scriptName = json(object(
                field("type", "text/javascript"),
                field("source", "require('roles/conditionalRoles').evaluateConditionalRoles(user, rolesPropName, " +
                        "existingConditionalRoles, userRoleGrants);")
        ));
        ScriptEntry scriptEntry = getScriptRegistry().takeScript(scriptName);
        assertThat(scriptEntry).isNotNull();
        Script script = scriptEntry.getScript(new RootContext());
        script.put("user", user);
        script.put("rolesPropName", ROLES);
        script.put("existingConditionalRoles", existingConditionalRoles);
        script.put("userRoleGrants", userRoleGrants);
        script.eval();
        assertThat(roleGrantsCorrect(user)).isTrue();
    }

    private static boolean userContainsRole(JsonValue user, String... roleRefs) {
        List<Map> roles = user.get(ROLES).asList(Map.class);
        for (Map role : roles) {
            for (String roleRef : roleRefs) {
                if (role.get(REF).toString().contains(roleRef)) {
                    return true;
                }
            }
        }
        return false;
    }

    private JsonValue baseUser() {
        int ndx = 5;
        return json(object(
                field(ID, "user" + ndx),
                field(REV, "1"),
                field(MAIL, "user" + ndx + "@example.com"),
                field(GIVEN_NAME, "fooUser" + ndx),
                field(SN, "user" + ndx + "sn"),
                field(DESCRIPTION, "description of fooUser " + ndx),
                field(USER_NAME, "user" + ndx + "@example.com"),
                field(TELEPHONE, "1234" + ndx),
                field(STATUS, "active")
        ));
    }

    private JsonValue userWithRoleGrant(Role role, GrantType grantType) {
        JsonValue baseUser = baseUser();
        if (Role.NONE != role) {
            baseUser.put(ROLES, getRoleGrant(role, grantType));
        }
        return baseUser;
    }

    private Map<String, Object> getRoleGrant(Role role, GrantType grantType) {
        return json(object(
                field(REF, "managed/role/" + role.name()),
                field(REF_PROPERTIES, getRoleGrantRefProperties(grantType)
        ))).asMap();
    }

    private Map<String, Object> getRoleGrantRefProperties(GrantType grantType) {
        JsonValue refProperties = json(object(
                field(REV, 1),
                field(ID, ROLE_GRANT_ID)));
        if (GrantType.CONDITIONAL == grantType) {
            refProperties.add(GRANT_TYPE, CONDITIONAL);
        }
        return refProperties.asMap();
    }

    private Map<String, Object> getRole(Role role) {
        return json(object(
                field(ID, role.name()),
                field(REV, "1"),
                field(NAME, "role " + role.name()),
                field(CONDITION, role.getRoleCondition()),
                field(DESCRIPTION, role.getRoleDescription()))).asMap();
    }

    private boolean roleGrantsCorrect(JsonValue user) {
        return userContainsRole(user, Role.MAIL_CO_COM.name(), Role.MAIL_CO_EXAMPLE.name()) &&
                !userContainsRole(user, Role.MAIL_CO_GOOGLE.name()) ;
    }

    private JsonValue getUserRoleGrants(Role conditionalRole, Role directRole) {
        return json(object(
                field(CONDITIONAL_GRANTS, conditionalRole != null ?
                        Collections.singletonList(getRoleGrant(conditionalRole, GrantType.CONDITIONAL)) : Collections.emptyList()),
                field(DIRECT_GRANTS, directRole != null ?
                        Collections.singletonList(getRoleGrant(directRole, GrantType.DIRECT)) : Collections.emptyList())));
    }

    private JsonValue systemConditionalRoles() {
        List<Map<String, Object>> roles = new ArrayList<>(3);
        roles.add(getRole(Role.MAIL_CO_COM));
        roles.add(getRole(Role.MAIL_CO_EXAMPLE));
        roles.add(getRole(Role.MAIL_CO_GOOGLE));
        return new JsonValue(roles);
    }
}

File: openidm-zip/src/test/resources/bin/defaults/script/conditionalRolesTest.js
/*
 * The contents of this file are subject to the terms of the Common Development and
 * Distribution License (the License). You may not use this file except in compliance with the
 * License.
 *
 * You can obtain a copy of the License at legal/CDDLv1.0.txt. See the License for the
 * specific language governing permission and limitations under the License.
 *
 * When distributing Covered Software, include this CDDL Header Notice in each file and include
 * the License file at legal/CDDLv1.0.txt. If applicable, add the following below the CDDL
 * Header, with the fields enclosed by brackets [] replaced by your own identifying
 * information: "Portions copyright [year] [name of copyright owner]".
 *
 * Copyright 2016 ForgeRock AS.
 */

/**
 * Tests against conditionalRoles.js.
 */
exports.test = function() {
    var conditionalRoles = require("roles/conditionalRoles");
    var _ = require('lib/lodash');

    isTemporalConstraintsMultiValue();

    function isTemporalConstraintsMultiValue() {
            var dateUtil = org.forgerock.openidm.util.DateUtil.getDateUtil(),
                now = dateUtil.currentDateTime(),
                currentDuration = dateUtil.formatDateTime(now.minusDays(1))
                    + "/" + dateUtil.formatDateTime(now.plusDays(1)),
                expiredDuration = dateUtil.formatDateTime(now.minusDays(2))
                    + "/" + dateUtil.formatDateTime(now.minusDays(1));
            [
                [
                    {
                        "_id": "roleWithOneConstraint",
                        "temporalConstraints": [
                            {
                                "duration": currentDuration
                            }
                        ]
                    },
                    false
                ],
                [
                    {
                        "_id": "roleWithMultipleConstraints",
                        "temporalConstraints": [
                            {
                                "duration": currentDuration
                            },
                            {
                                "duration": expiredDuration
                            }
                        ]
                    },
                    true
                ]
            ].map(
                function (testcase) {
                    (function (role, expectedResult) {
                        var result = conditionalRoles.isTemporalConstraintsMultiValue(role)
                        if (result !== expectedResult) {
                            throw {
                                "message": "Determining if temporalConstraints is multivalue " + JSON.stringify(role) +
                                ", got <" + result + ">, expected <" + expectedResult + ">"
                            };
                        }
                    }).apply(null, testcase);
                });
        }

    evaluateConditionalRoles();
    function evaluateConditionalRoles() {
        var
            mailGoogleRoleDef = {
                'name' : 'MailGoogleRole',
                'condition' : "/mail co 'google'",
                'description' : 'does mail contain google'
            },
            mailExampleRoleDef = {
                'name' : 'MailExampleRole',
                'condition' : "/mail co 'example'",
                'description' : 'does mail contain example'
            },
            mailComRoleDef = {
                'name' : 'MailComRole',
                'condition' : "/mail co 'com'",
                'description' : 'does mail contain com'
            };

        function getSystemConditionalRoles() {
            return [
                getRole(mailGoogleRoleDef),
                getRole(mailExampleRoleDef),
                getRole(mailComRoleDef)
            ];
        }
        
        function getRole(roleDef) {
            return {
                '_id' : roleDef.name,
                '_rev' : '1',
                'name' :  'role' + roleDef.name,
                'condition' :  roleDef.condition,
                'description' : roleDef.description
            };
        }

        function baseUser(userName, email) {
            var userId = 'user' + userName;
            return {
                '_id' : userId,
                '_ref' : '1',
                //passing in email as param - cannot catenate the userId with the mail suffix, as this will cause condition evaluation to fail
                //due to mozilla type marshalling - strings composed with + marshal to org.mozilla.javascript.ConsString
                //which will fail the QueryFilterCondition#isCompatible type congruence check.
                'mail' : email,
                'givenName' : userId + ' givenName',
                'sn' : userId + 'sn',
                'description' : 'description of ' + userId,
                'userName' : userId + '@example.com',
                'status' : 'active'
            };
        }

        function getRoleGrant(roleDef, isConditional) {
            var grantDef = {}
            if (roleDef) {
                grantDef = {
                    '_ref': 'managed/role/' + roleDef.name,
                    '_refProperties': getRoleGrantRefProperties(isConditional)
                };
            }
            return grantDef;
        }

        function getRoleGrantRefProperties(isConditional) {
            var refProperties =
            {
                '_rev': '1',
                '_id': 'superRandomGrantId'
            }
            if (isConditional) {
                    refProperties['_grantType'] = 'conditional';
            }
            return refProperties;
        }

        function getUserRoleGrants(conditionalRole, directRole) {
            return {
                'conditionalGrants' : conditionalRole != null ? [getRoleGrant(conditionalRole, true)] : [],
                'directGrants' : directRole != null ? [getRoleGrant(directRole, false)] : []
            };
        }

        function userHasGrants(userRoles, roleIds) {
            var roleIdInGrants =
                function(roleId) {
                    return _.find(userRoles, function (roleGrant) { return 'managed/role/' + roleId === roleGrant._ref}) !== undefined;
                };
            return roleIds.every(roleIdInGrants)
        }

        /*
        The conditionalRoles script does not create an id for the grants it creates - this occurs on persistence. So
        if the grant was created by the script, it will have an undefined id.
         */
        function grantsCreatedByScriptInvocation(userRoles, roleIds) {
            var roleIdInGrants =
                function(roleId) {
                    var grant =  _.find(userRoles, function (roleGrant) { return 'managed/role/' + roleId === roleGrant._ref});
                    if (grant === undefined) {
                        return false;
                    }
                    return grant._refProperties._id === undefined;
                };
            return roleIds.every(roleIdInGrants)
        }

        [
            [
                baseUser('googleUser', 'baseUser@google.com'),
                getSystemConditionalRoles(),
                getUserRoleGrants(null, mailGoogleRoleDef),
                [mailComRoleDef.name, mailGoogleRoleDef.name],
                [mailComRoleDef.name]
            ],
            [
                baseUser('exampleUser', 'baseUser@example.com'),
                getSystemConditionalRoles(),
                getUserRoleGrants(null, null),
                [mailComRoleDef.name, mailExampleRoleDef.name],
                [mailComRoleDef.name, mailExampleRoleDef.name]
            ],
            [
                baseUser('exampleUser', 'baseUser@dillrod.com'),
                getSystemConditionalRoles(),
                getUserRoleGrants(null, null),
                [mailComRoleDef.name],
                [mailComRoleDef.name]
            ],
            [
                baseUser('exampleUser', 'baseUser@google.com'),
                getSystemConditionalRoles(),
                getUserRoleGrants(mailGoogleRoleDef, null),
                [mailComRoleDef.name, mailGoogleRoleDef.name],
                [mailComRoleDef.name]
            ]

        ].map(
            function (testcase) {
                (function (user, systemConditionalRoles, userRoleGrants, expectedRoleIds, createdRoleIds) {
                    conditionalRoles.evaluateConditionalRoles(user, 'roles', systemConditionalRoles, userRoleGrants);
                    if (!userHasGrants(user.roles, expectedRoleIds) || !grantsCreatedByScriptInvocation(user.roles, createdRoleIds)) {
                        throw {
                            "message": "Evaluating conditional roles failure: script-produced user roles: <" + JSON.stringify(user.roles) +
                            ">, but should have had grants corresponding to role ids <" + JSON.stringify(expectedRoleIds) + ">" +
                            "of which <" + JSON.stringify(createdRoleIds) + "> role ids should have been created by the script."
                        };
                    }
                }).apply(null, testcase);
            });

    }
}